강원대_안드로이드_이아림_2주차 과제_제출

arieum opened the pull request:
- 어려웠던 점
0. 1단계를 먼저하고 넘어갔어야 했는데 2단계 먼저 구현하는 바람에 중간에 mvvm 패턴으로 바꾸는 데 실패했습니다😢 
mvvm이 뭔지도 모른 상태로 일단 다 구현해놓고 나중에 mvvm을 구현할 때에는 새 클래스들(view, model, viewmodel)을 만들고 옮겨넣기만 하면 되는 줄 알았는데 전혀 아니었습니다.
2. MVVM을 직접 구현할 때 plugins kapt 추가가 제대로 진행되지 않았습니다
3. 코드를 짜다보니 MVVM이 왜 필요한지는 알 것 같습니다. 지금 제 코드는 클래스 간에 의존성(결합성?)이 너무 강해서 한 줄을 수정할 때 여러 클래스를 오가야 했습니다😨
4. 1차 피드백때 `notifyDataSetChanged()` 권장하지 않는다는 피드백을 보고 `notifyItemInserted(researchList.size - 1)`로 수정할 수 있었습니다. 하지만 `resultAdapter.notifyDataSetChanged()` 이 부분은 클래스 간 의존성때문인지 아무리 수정해도 원하는 기능대로 되지않아 이 코드대로 방치했습니다..

arieum committed:
feat: Add basic constructure for README.md

arieum committed:
Update README.md

arieum committed:
feat: Add initial_search_screen UI layout(activity_main.xml)

arieum committed:
feat: Add tab_card.xml for displaying search_results

arieum committed:
feat: Define Place_data_class with properties(img, name, location, category)

arieum committed:
feat: Match variables between activity_main.xml and MainActivity

arieum committed:
feat: Create place_cardview.xml for Recyclerview of serch results

arieum committed:
feat: Add PlaceContract.kt

arieum committed:
refact: Regenerate ⓒPlace with properties(img, name, location, category)

arieum committed:
feat: Add PlaceDbHelper.kt

arieum committed:
feat: Add dbHelper for database_management in MainActivity

arieum committed:
feat: Add PlaceRepository.kt to insert data in database

arieum committed:
feat: Add table_Myresearch to record myresearch_log

arieum committed:
feat: Insert initial_data into table_place from MainActivity

arieum committed:
Create RecyclerViewAdapter

arieum committed:
feat: Implement dynamic filtering of ReclyerView based on input_EditView

arieum committed:
feat: Add cardView_setOnClickListener for recording log_database

arieum committed:
feat: Create myresearch_table in MyPlace database

arieum committed:
fix: Error

arieum committed:
fix: Error2

arieum committed:
feat: Implement TapViewAdapter to display search_History

arieum committed:
feat: Add functionality to display clicked items in a tab_recyclerview

arieum committed:
style: Update tab_card.xml design

arieum committed:
feat: Add functionality to check if a place_object already exists in research

arieum committed:
feat: Override equals_method in Place_Class

arieum committed:
feat: Override hashCode_method in Place_class to maintain consistency with equals_method

arieum committed:
feat: Prevent duplicate Place_entries in research_table db

arieum committed:
feat: Add hasReasearchEntries_method to check if the research_table contains any entries

arieum committed:
feat: Display research_entries in tableRecyclerView and update Visibility

arieum committed:
feat: Add deleteResearchEntry_method to PlaceRepository to delete entries from research_table

arieum committed:
feat: Implement tab_item_Removal from research_list and research_table on cacelButton_click

- Updated TapViewAdapter to  handle cancelButton click events
- Modified MainActivity to pass placeRepository to TapviewAdapter for
  db_operation
- TabRecyclerview_visibility is updated after item is removed

arieum committed:
feat: Minimize layout_depth by changing Total_layout to ConstringLayout in activity_main.xml

arieum committed:
fix: Change layout_width from 0dp to match_parent to ensure fixed width in Constriaint_layout

arieum committed:
chore: Remove unnecessary comments

arieum committed:
feat: Add search functionality to filter place by name or category

arieum committed:
feat: Optimize RecyclerView updates by optionally removing, adding, updating items in place_list to replace notifyDataSetChanged()

arieum committed:
revert: Revert to using notifyDataSetChanged() due to issues that updatePlaceList_method

arieum committed:
feat: Add functionality to move clicked_item to the end of the list

arieum committed:
revert: Remove functionality to move clicked_item

arieum committed:
Add files via upload

arieum committed:
Update README.md

arieum committed:
fix: Clear the list when the input_form is empty

arieum committed:
Merge branch 'arieum_step1' of https://github.com/arieum/android-map-keyword into arieum_step1

mkSpace commented on app/src/main/java/campus/tech/kakao/map/MainActivity.kt:
```diff
@@ -1,11 +1,105 @@
+
+        updateTabRecyclerViewVisibility()
+
+        input.addTextChangedListener(object : TextWatcher {
+            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
+            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
+            override fun afterTextChanged(s: Editable?) {
+                filterList(s.toString())
+            }
+        })
```
<img width="664" alt="image" src="https://github.com/kakao-tech-campus-2nd-step2/android-map-keyword/assets/24197741/4637b831-1be6-4266-9163-11616c38cc1e">

addTextChangedListener 확장함수를 사용하면 원하시는 부분만 재정의 가능합니다 이걸 활용해보세요!
그리고 TextWatcher를 등록했다면 View가 삭제될 때 removeTextChangedListener를 통해 삭제해주세요!

mkSpace commented on app/src/main/java/campus/tech/kakao/map/Place.kt:
```diff
@@ -0,0 +1,31 @@
+    val location: String,
+    val category: String
+) {
+    override fun equals(other: Any?): Boolean {
```
equals와 hashCode를 재정의하신 이유가 있을까요?

arieum replied:
처음에 contains를 사용했을 때 계속 빨간줄이 쳐져있었습니다. 구글링해봤더니 equals와 hashcode를 재정의해야 알맞게 객체를 비교할 수 있는 줄 알고 재정의했었던 것 같습니다..! 근데 코치님 리뷰보고 지워봤더니 없어도 실행가능하네요..😅

mkSpace commented on app/src/main/java/campus/tech/kakao/map/PlaceDbHelper.kt:
```diff
@@ -0,0 +1,47 @@
+    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
+        db?.execSQL(SQL_DELETE_PLACE)
+        db?.execSQL(SQL_DELETE_RESEARCH)
+        onCreate(db)
```
onCreate는 콜백메서드이니 직접 호출하는것보단 필요한 메서드를 따로 빼서 정의하시는게 좋을 것 같습니다

mkSpace commented on app/src/main/java/campus/tech/kakao/map/PlaceRepository.kt:
```diff
@@ -0,0 +1,140 @@
+import android.provider.BaseColumns
+import android.util.Log
+
+class PlaceRepository(context: MainActivity) {
```
Context가 아닌 Activity를 받는 이유가 있을까요?

arieum replied:
"나중에 여기 MainAcitivty 넣어줘야지" 생각하면서 코드 작성하느라 저도 모르게 Context가 아니라 MainActivity로 작성한 것 같습니다😢

mkSpace commented on app/src/main/java/campus/tech/kakao/map/PlaceRepository.kt:
```diff
@@ -0,0 +1,140 @@
+                    Log.d("PlaceRepository", "Successfully inserted row for ${place.name}")
+                }
+            }
+            cursor.close()
```
위에서 예외가 발생하면 이 코드가 실행되지 않을 수 있습니다. finally에서 처리해주세요!

mkSpace commented on app/src/main/java/campus/tech/kakao/map/PlaceRepository.kt:
```diff
@@ -0,0 +1,140 @@
+        val db = dbHelper.readableDatabase
+        val cursor = db.rawQuery("SELECT COUNT(*) FROM ${MyPlaceContract.Research.TABLE_NAME}", null)
+        return if (cursor.moveToFirst()) {
+            val count = cursor.getInt(0)
```
```suggestion
            val count = cursor.getIntOrNull(0)
```
getIntOrNull을 사용해서 null-safe 하게 처리하면 좋겠네요

mkSpace commented on app/src/main/java/campus/tech/kakao/map/PlaceRepository.kt:
```diff
@@ -0,0 +1,140 @@
+    private var placeList = mutableListOf<Place>()
+
+    init {
+        dbHelper.writableDatabase
```
이건 왜 작성하신건가요?

arieum replied:
초반에 DB가 익숙하지 않을때 테이블이 생성되지 않아 이것저것 써본 코드들을 다 지웠다고 생각했는데 남아있는 코드들이 있는 것 같습니다..ㅎ

mkSpace commented on app/src/main/java/campus/tech/kakao/map/PlaceRepository.kt:
```diff
@@ -0,0 +1,140 @@
+import android.util.Log
+
+class PlaceRepository(context: MainActivity) {
+    private val dbHelper = PlaceDbHelper(context)
```
dbHelper로 db를 open 한 로직만 보이네요. 어디에서 close 해야 할까요?

mkSpace commented on app/src/main/java/campus/tech/kakao/map/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,61 @@
+
+class RecyclerViewAdapter(
+    var placeList: MutableList<Place>,
+    var inflater: LayoutInflater,
```
외부에서 inflater를 받아서 사용하지 않고 활용해보세요. onCreateViewHolder에서 parent를 통해 LayoutInflater를 사용할 수 있습니다.

mkSpace commented on app/src/main/java/campus/tech/kakao/map/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,61 @@
+class RecyclerViewAdapter(
+    var placeList: MutableList<Place>,
+    var inflater: LayoutInflater,
+    var placeRepository: PlaceRepository
```
Adapter가 Repository까지 알면 의존성이 매우 강결합 됩니다.
로직은 항상 외부에 있게 작성해주세요. Adapter는 콜백 메서드만 받아서 처리하는게 이상적입니다.

mkSpace commented on app/src/main/java/campus/tech/kakao/map/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,61 @@
+    var placeList: MutableList<Place>,
+    var inflater: LayoutInflater,
+    var placeRepository: PlaceRepository
+) : RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder>() {
```
기본 Adapter를 구현하는것도 좋지만 시간이 괜찮으시면 ListAdapter + Diff 를 사용해서 리팩토링 해보시겠어요?

mkSpace commented on app/src/main/java/campus/tech/kakao/map/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,61 @@
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        val place: Place = placeList.get(position)
+
+        val img = when (place.category) {
+            "카페" -> R.drawable.cafe
+            "약국" -> R.drawable.hospital
+            else -> R.drawable.location
```
String을 직접 작성하시는경우 Human-Error가 발생하기 쉽습니다. enum class를 활요해보시겠어요?

mkSpace commented on app/src/main/java/campus/tech/kakao/map/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,61 @@
+            location = itemView.findViewById(R.id.place_location)
+            category = itemView.findViewById(R.id.place_category)
+
+            itemView.setOnClickListener {
```
ViewHolder의 init 블록에서 리스너를 등록해주시는 코드 잘 작성해주셨습니다! 아래에 로직 부분만 외부에서 받아서 실행하는걸로 변경해보면 좋을 것 같아요

mkSpace reviewed (changes requested):
이번 Step도 고생많으셨습니다!

아직 MVVM 아키텍처가 적용되지 않아서 그 부분은 리뷰하지 못했네요.

MVVM을 당장 적용하기 힘드시다면 좀 더 간단한 Toy 프로젝트를 만들어서 공부해보시는건 어떨까요?

리뷰 남겨놓으니 의견 남겨주세요!

arieum committed:
feat: Add EditView_listener_extension function for TextWater

1. Create EditViewAddTextChangedListner.extension file
2. Use extension function
3. Remove TextWater on 'onDestroy'

arieum committed:
refactor: Remove hashcode&equals_override_method

arieum committed:
feat: Create PlaceCategory_enum_class

arieum committed:
Refactor: USE enum_class and UPDATE relvant_methodto prevent HUMAN-ERROR

arieum committed:
refactor: Ensure cursor.close() is called in finally_block

arieum committed:
refactor: Ensure db & cursor_close() using finally blocks

arieum committed:
refactor: Use LayoutInflater from context of parent instead of passing it externally

arieum committed:
refactor: Remove PlaceRepository dependency from RecyclerViewAdapter and handle item_Click_events in MainActivity

arieum committed:
refactor: Remove PlaceRepository depency from TabRecylcerVIewAdapter and handle item_click_event in MainActivty by using method_removeResearchList

arieum committed:
Create: Add PlaceDiffCallBack class for ListAdapter

arieum commented:
- 1차 피드백 이후
1. 저도 처음에 코드 작성하면서 PlaceRepository에 의존하는 코드들이 너무 많다고 생각했지만 어떻게 수정해야하는지 감이 안왔습니다. 코치님 피드백 덕분에 코드가 처음보다는 깔끔해진 것 같습니다 (물론 더 좋은 가독성을 위해서는 더 연습해야할  것 같습니다.) 
2. 카테캠 1차교육에서 기본적으로 RecyclerViewAdapter 상속은 `RecyclerView.Adapter`으로 배웠는데 코치님의 피드백대로 `ListAdapter<>` 상속에 대해 알고나서 먼저 미니프로젝트에 테스트구현중입니다! 😊
